File: Java/프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        HashSet<String> hs = new HashSet<>(Arrays.asList(gems));
        int gemsCount = hs.size();
        
        HashMap<String, Integer> hm = new HashMap<>();
        int[] answer = new int[2];
        int start = 0;
        int min = Integer.MAX_VALUE;
        
        for (int end = 0; end < gems.length; end++) {
            String lastGem = gems[end];
            hm.put(lastGem, hm.getOrDefault(lastGem, 0) + 1);
        
            while (hm.size() == gemsCount) {
                if (end - start < min) {
                    min = end - start;              
                    answer[0] = start + 1;
                    answer[1] = end + 1;
                }
                
                String firstGem = gems[start];
                hm.put(firstGem, hm.getOrDefault(firstGem, 0) - 1);
                if (hm.getOrDefault(firstGem, 0) <= 0) {
                    hm.remove(firstGem);
                }
                
                start++;
            }
        }
        
        return answer;
    }
}

File: Swift/프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.swift
import Foundation

func solution(_ gems:[String]) -> [Int] {
    var set: Set<String> = Set(gems)
    let gemsCount = set.count
    
    var dic: [String: Int] = [:]
    var answer: [Int] = [0, 0]
    var start = 0
    var min = Int.max
    
    for end in 0..<gems.count {
        let endGem = gems[end]
        dic[endGem] = (dic[endGem] ?? 0) + 1
        
        while dic.count == gemsCount {
            if (end - start < min) {
                min = end - start
                answer[0] = start + 1
                answer[1] = end + 1
            }
            
            let firstGem = gems[start]
            dic[firstGem] = (dic[firstGem] ?? 0) - 1
            if (dic[firstGem] ?? 0) <= 0 {
                dic.removeValue(forKey: firstGem)
            }
            
            start += 1
        }
    }
    
    return answer
}